core/controllers/reader.py

```diff
@@ -122,31 +123,67 @@ def classify(exp_id, state, answer, params):
-            best_matched_truth_value = ored_truth_value
-            best_matched_rule_spec = best_rule_spec
-            best_matched_answer_group = answer_group
-            best_matched_answer_group_index = i
+            if rule_spec.rule_type != rule_domain.FUZZY_RULE_TYPE:
+                evaluated_truth_value = rule_domain.evaluate_rule(
+                    rule_spec, input_type, params, normalized_answer, fs)
+                if evaluated_truth_value > best_matched_truth_value:
+                    best_matched_truth_value = evaluated_truth_value
+                    best_matched_rule_spec = rule_spec
+                    best_matched_answer_group = answer_group
+                    best_matched_answer_group_index = i
+
+    # Find the maximum soft rule that matches. This is
+    # done by ORing (maximizing) all truth values of all rules over all answer
+    # groups. The group with the highest truth value is considered the best
+    # match.
+    if best_matched_answer_group is None:
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    evaluated_truth_value = rule_domain.evaluate_rule(
+                        rule_spec, input_type, params, normalized_answer, fs)
+                    if evaluated_truth_value > best_matched_truth_value:
+                        best_matched_truth_value = evaluated_truth_value
+                        best_matched_rule_spec = rule_spec
+                        best_matched_answer_group = answer_group
+                        best_matched_answer_group_index = i
+
+    # Run the classifier
+    if best_matched_answer_group is None:
+        sc = classifier_services.StringClassifier()
+        training_examples = [[doc, []] for doc in
+            state.interaction.confirmed_unclassified_answers]
+        for (i, answer_group) in enumerate(state.interaction.answer_groups):
+            for rule_spec in answer_group.rule_specs:
+                if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
+                    training_examples.extend([[doc, [str(i)]] for doc in
+                        rule_spec.inputs['training_data']])
+        if len(training_examples) > 0:
+            sc.load_examples(training_examples)
+            doc_ids = sc.add_examples_for_predicting([answer])
+            predicted_label = sc.predict_label_for_doc(doc_ids[0])
+            if predicted_label != '_default':
+                predicted_answer_group_index = int(predicted_label)
+                predicted_answer_group = state.interaction.answer_groups[
+                    predicted_answer_group_index]
+                best_matched_truth_value = rule_domain.CERTAIN_TRUE_VALUE
+                for rule_spec in answer_group.rule_specs:
+                    if rule_spec.rule_type == rule_domain.FUZZY_RULE_TYPE:
```

Comment:
I might've heard from Sean, but I thought each answer group only had at most one fuzzy rule. Looks like that's not true. Or I may be getting terminology incorrect. What does it mean to have more than one though? If fuzzy rules are based off "is similar to" then having two corresponding to an answer group seems the same has having one with the union of the two.

Acknowledging your comments, Let me know about the above and I'll rethink this part of the code.